lib/xdrgen/generators/go.rb

```diff
@@ -336,6 +419,126 @@ def render_binary_interface(out, name)
         out.break
       end
 
+      # render_encode_to_body assumes there is an `e` variable containing an
+      # xdr.Encoder, and a variable defined by `name` that is the value to
+      # encode.
+      def render_encode_to_body(out, var, type, self_encode:)
+        optional = type.sub_type == :optional
+        if optional
+          out.puts "  _, err = e.EncodeBool(#{var} != nil)"
+          out.puts "  if err != nil {"
+          out.puts "    return err"
+          out.puts "  }"
+          out.puts "  if #{var} != nil {"
+          var = "(*#{var})"
+        end
+        case type
+        when AST::Typespecs::UnsignedHyper
+          out.puts "  _, err = e.EncodeUhyper(uint64(#{var}))"
+        when AST::Typespecs::Hyper
+          out.puts "  _, err = e.EncodeHyper(int64(#{var}))"
+        when AST::Typespecs::UnsignedInt
+          out.puts "  _, err = e.EncodeUint(uint32(#{var}))"
+        when AST::Typespecs::Int, AST::Definitions::Enum
+          out.puts "  _, err = e.EncodeInt(int32(#{var}))"
+        when AST::Typespecs::String
+          out.puts "  _, err = e.EncodeString(string(#{var}))"
+        when AST::Typespecs::Opaque
+          if type.fixed?
+            out.puts "  _, err = e.EncodeFixedOpaque(#{var}[:])"
+          else
+            out.puts "  _, err = e.EncodeOpaque(#{var}[:])"
+          end
+        when AST::Typespecs::Simple
+          case type.sub_type
+          when :simple, :optional
+            optional_within = type.is_a?(AST::Identifier) && type.resolved_type.sub_type == :optional
+            if optional_within
+              out.puts "  _, err = e.EncodeBool(#{var} != nil)"
+              out.puts "  if err != nil {"
+              out.puts "    return err"
+              out.puts "  }"
+              out.puts "  if #{var} != nil {"
+              var = "(*#{var})"
+            end
```

Comment:
If Go supports pointer receiver base types on method declarations (golang/go#48592) in the future then this logic could live inside the `EncodeTo` of that type, which would be simpler, rather than having to be lifted up to here and the array sections below.